### PR TITLE
feat: rest sink compress

### PR DESCRIPTION
**Pull request by carlclone**

### CODE REVIEW

**Comment by carlclone** [comment]
#2714

File: internal/io/http/rest_sink_test.go [hunk 1]
```diff
@@ -149,7 +154,8 @@ func TestRestSink_Apply(t *testing.T) {
 				ContentType: "application/x-www-form-urlencoded;param=value",
 				Body:        `ab=hello2`,
 			}},
-		}, {
+		},
```
Comment (wfnuser) [inline internal/io/http/rest_sink_test.go:157]: Are these lint format results or just some irrelevant commits?

File: internal/io/http/rest_sink.go [hunk 2]
```diff
@@ -16,21 +16,23 @@ package http
...
 	"fmt"
+	"github.com/pingcap/failpoint"
 	"net/http"
 	"net/url"
 	"strings"
 	"time"
 
-	"github.com/pingcap/failpoint"
-
+	"github.com/lf-edge/ekuiper/internal/compressor"
```
Comment (wfnuser) [inline internal/io/http/rest_sink.go:25]: I guess there are some lint problems here.

File: internal/io/http/rest_sink.go [hunk 3]
```diff
@@ -40,7 +42,17 @@ func (ms *RestSink) Validate(props map[string]interface{}) error {
...
 func (ms *RestSink) Configure(ps map[string]interface{}) error {
 	conf.Log.Infof("Initialized rest sink with configurations %#v.", ps)
-	return ms.Validate(ps)
+	err := ms.Validate(ps)
+	if err != nil {
+		return err
+	}
+	if ms.config.Compression != "" {
+		ms.compressor, err = compressor.GetCompressor(ms.config.Compression)
+		if err != nil {
```
Comment (wfnuser) [inline internal/io/http/rest_sink.go:51]: I guess the compressor.GetCompressor error message is clear enough. Though, no harm to make it more specific.

File: internal/io/http/rest_sink_test.go [hunk 4]
```diff
@@ -565,3 +597,9 @@ func TestRestSinkMethod(t *testing.T) {
 	err := rs.Validate(map[string]interface{}{"method": "head"})
 	require.Error(t, err)
 }
+
+func TestRestSinkConfigure(t *testing.T) {
```
Comment (wfnuser) [inline internal/io/http/rest_sink_test.go:601]: Better to make this test name clearly showing it's a failure test. I guess the intention is to test unsupported compression method.

File: internal/io/http/rest_sink.go (hunk 2 above)
Comment (carlclone) [reply internal/io/http/rest_sink.go:26]: thanks , fixed

File: internal/io/http/rest_sink.go (hunk 3 above)
Comment (carlclone) [reply internal/io/http/rest_sink.go]: fixed

File: internal/io/http/rest_sink_test.go (hunk 4 above)
Comment (carlclone) [reply internal/io/http/rest_sink_test.go]: fixed

**Review by wfnuser — APPROVED** [review]

**Comment by ngjaying** [comment]
@carlclone The FVT tests fail. Do you need help on that?

File: internal/io/http/rest_sink.go [hunk 5]
```diff
@@ -64,6 +79,13 @@ func (ms *RestSink) collectWithUrl(ctx api.StreamContext, item interface{}, desU
 		return fmt.Errorf("rest sink decode data error: %v", err)
 	}
 
+	if ms.compressor != nil {
+		decodedData, err = ms.compressor.Compress(decodedData)
+		if err != nil {
+			return err
+		}
+	}
+
```
Comment (ngjaying) [inline internal/io/http/rest_sink.go:88]: Http header needs to be set when using compression like `req.Header.Set("Content-Encoding", "gzip")`

**Review by ngjaying — CHANGES_REQUESTED** [review]

File: internal/io/http/rest_sink.go (hunk 5 above)
Comment (carlclone) [reply internal/io/http/rest_sink.go:88]: the Content-Encoding header seems to only set in the response rather than request

**Comment by carlclone** [comment]
will create a new pr , this may not being considerate enough

**Pull request closed without merging**